test/tls12groupselection_test.c

```diff
@@ -0,0 +1,201 @@
+#include "testutil.h"
+#include "helpers/ssltestlib.h"
+#include <openssl/objects.h>
+
+#define TEST_true_or_end(a) if (!TEST_true(a)) \
+        goto end;
+
+#define SERVER_PREFERENCE 1
+#define CLIENT_PREFERENCE 0
+
+#define WORK_ON_SSL_OBJECT 1
+#define WORK_ON_CONTEXT 0
+
+#if !defined(OPENSSL_NO_EC) && !defined(OPENSSL_NO_TLS1_2)
+
+static char *cert = NULL;
+static char *privkey = NULL;
+
+struct tls12groupselection_test_st {
+    const char *client_groups;
+    const char *server_groups;
+    int preference;
+    const char *expected_group;
+};
+
+static const struct tls12groupselection_test_st tls12groupselection_tests[] =
+    {
+        { "secp521r1:secp384r1:X25519:prime256v1:X448",  /* test 0 */
+          "X25519:secp521r1:secp384r1:prime256v1:X448",
+          CLIENT_PREFERENCE,
+          "secp521r1"
+        },
+        { "secp521r1:secp384r1:X25519:prime256v1:X448",  /* test 1 */
+          "X25519:secp521r1:secp384r1:prime256v1:X448",
+          SERVER_PREFERENCE,
+          "X25519"
+        },
+        { "secp384r1:prime256v1:X448",                   /* test 2 */
+          "prime256v1:X448:secp384r1",
+          CLIENT_PREFERENCE,
+          "secp384r1"
+        },
+        { "secp384r1:prime256v1:X448",                   /* test 3 */
+          "prime256v1:X448:secp384r1",
+          SERVER_PREFERENCE,
+          "prime256v1"
+        },
+        { "secp521r1:secp384r1:X25519:prime256v1:X448",  /* test 4 */
+          "X25519:secp384r1:prime256v1",
+          CLIENT_PREFERENCE,
+          "secp384r1"
+        },
+        { "secp521r1:secp384r1:X25519:prime256v1:X448", /* test 5 */
+          "X25519:secp384r1:prime256v1",
+          SERVER_PREFERENCE,
+          "X25519"
+        },
+        { NULL,                                        /* test 6 */
+          NULL,
+          CLIENT_PREFERENCE,
+          "X25519"
+        },
+        { NULL,                                        /* test 7 */
+          NULL,
+          SERVER_PREFERENCE,
+          "X25519"
+        },
+        { "secp521r1:secp384r1:X25519:prime256v1", /* test 8 */
+          "X448:X25519:secp384r1:prime256v1",
+          SERVER_PREFERENCE,
+          "X25519"
+        },
```

Comment:
Should we also have some tests using the new TLSv1.3 syntax but with TLSv1.2?
Probably also a test where there is no overlap?